keras/engine/training.py

```diff
@@ -1985,21 +1896,23 @@ def evaluate_generator(self, generator, steps,
         wait_time = 0.01
         all_outs = []
         batch_sizes = []
+        is_dataset = isinstance(generator, Dataset)
+        if not is_dataset and pickle_safe:
+            warnings.warn(
+                """Using a generator with `pickle_safe=True` may duplicate your data.
+                Please considers using the `keras.data.Dataset` object.""")
         enqueuer = None
 
         try:
-            enqueuer = GeneratorEnqueuer(generator, pickle_safe=pickle_safe)
+            if is_dataset:
+                enqueuer = OrderedEnqueuer(generator, pickle_safe=pickle_safe)
+            else:
+                enqueuer = GeneratorEnqueuer(generator, pickle_safe=pickle_safe, wait_time=wait_time)
             enqueuer.start(workers=workers, max_q_size=max_q_size)
+            output_generator = enqueuer.get()
 
             while steps_done < steps:
-                generator_output = None
-                while enqueuer.is_running():
-                    if not enqueuer.queue.empty():
-                        generator_output = enqueuer.queue.get()
-                        break
-                    else:
-                        time.sleep(wait_time)
-
+                generator_output = next(output_generator)
```

Comment:
This is done here : https://github.com/fchollet/keras/pull/6891/files/1c38bda3d3a7da31ce86fca72b3cf51a15b75642#diff-299cfd5886683a4b012f286403769fc1L2108